generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -164,6 +166,175 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by bitwise shifting left or right by a value
+   * between 0 and 8 bits, inclusive. For example, we can transform an integer i -> i >> 0 or
+   * i << 0. Generates an opaque value if the type of zero to be generated is an integer and the
+   * shading language in use supports bitwise operations.
+   *
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from bitwise shifting.
+   */
+  private Optional<Expr> opaqueZeroOrOneBitwiseShift(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    if (!BasicType.allIntegerTypes().contains(type)) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedBitwiseOperations()) {
+      return Optional.empty();
+    }
+    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
+    // The maximum safe value to shift zero without potentially causing undefined behavior is
+    // between zero and eight bits (as nine bits is the minimum precision of an integer in GLSL).
+    // If we're shifting one, we want to be able to shift backwards without losing information, and
+    // so our maximum shift value is seven bits.
+    final int bound = isZero ? 10 : 9;
+    final int maxValue = generator.nextInt(bound);
+    final Expr shiftValue = new FunctionCallExpr(
+        "clamp",
+        fuzzer.fuzzExpr(type, false, constContext, depth),
+        makeOpaqueZero(type, constContext, depth, fuzzer),
+        applyIdentityFunction(
+            new TypeConstructorExpr(
+                type.toString(),
+                type.getElementType() == BasicType.INT
+                    ? new IntConstantExpr(String.valueOf(maxValue))
+                    : new UIntConstantExpr(maxValue + "u")),
+            type, constContext, depth, fuzzer));
+    if (isZero) {
+      return Optional.of(
+          new ParenExpr(
+              new BinaryExpr(
+                  makeOpaqueZero(type, constContext, depth, fuzzer),
+                  applyIdentityFunction(shiftValue, type, constContext, depth, fuzzer),
+                  operator)));
+    } else {
+      assert !isZero;
+      assert operator == BinOp.SHL || operator == BinOp.SHR;
+      // We're going to shift twice in opposite directions by the same value.
```

Comment:
The only legal choice is to shift left then shift right.  The problem is that if we shift right first, the 1 will get shifted away and we'll end up with 0.  (I don't think it is worth allowing this for the special case where the shift value is 0.)

I suggest you choose the shift operator in the "isZero" branch, and then change this branch so that the order of operators is fixed.